c_src/quicer_config.c

```diff
@@ -2441,36 +2461,79 @@ get_str_from_map(ErlNifEnv *env,
   return enif_get_string(env, tmp_term, buff, tmp_len + 1, ERL_NIF_LATIN1);
 }
 
-BOOLEAN
-build_trustedstore(const char *cacertfile, X509_STORE **trusted_store)
+/*
+** Fill str_buffer with string value of key in map.
+** In case str_buffer is NULL, then new memory will be allocated,
+** and caller should free it after use.
+**
+** Returns NULL on error.
+*/
+char *
+str_from_map(ErlNifEnv *env,
+             ERL_NIF_TERM key,
+             const ERL_NIF_TERM *map,
+             char *str_buffer,
+             unsigned int max_len)
 {
-  X509_STORE *store = NULL;
-  X509_LOOKUP *lookup = NULL;
+  unsigned int len = 0;
+  ERL_NIF_TERM tmp_term;
 
-  if (cacertfile == NULL)
+  if (!enif_get_map_value(env, *map, key, &tmp_term))
     {
-      return FALSE;
+      goto exit;
     }
 
-  store = X509_STORE_new();
-  if (store == NULL)
+  if (ERL_NIF_TERM_TYPE_LIST != enif_term_type(env, tmp_term))
     {
-      return FALSE;
+      goto exit;
     }
 
-  lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file());
-  if (lookup == NULL)
+  if (!str_buffer && !enif_get_list_length(env, tmp_term, &len))
+    {
+      goto exit;
+    }
+  else
     {
-      X509_STORE_free(store);
-      return FALSE;
+      len = max_len;
     }
 
-  if (!X509_LOOKUP_load_file(lookup, cacertfile, X509_FILETYPE_PEM))
+  if (len == 0 || max_len < len)
     {
-      X509_STORE_free(store);
-      return FALSE;
+      goto exit;
+    }
+
+  if (!str_buffer)
+    {
+      str_buffer = (char *)malloc(len + 1);
+    }
+
+  if (enif_get_string(env, tmp_term, str_buffer, len + 1, ERL_NIF_LATIN1))
+    {
+      return str_buffer;
+    }
+
```

Comment:
Newbie question: if we allocated `str_buffer` but didn't return it above, should we free it here?

Reply:
yes. it should! thanks!